dali/pipeline/graph/cse.cc

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/pipeline/graph/cse.h"
+#include <functional>
+#include <map>
+#include <string>
+#include <utility>
+#include "dali/pipeline/dali.pb.h"
+
+namespace dali {
+namespace graph {
+
+namespace {
+
+/** Computes the CSE key by serializing the relevant subset of an OpSpec to protobuf */
+std::string OpSpecCSEKey(const OpSpec &spec) {
+  dali_proto::OpDef op;
+  op.set_name(spec.SchemaName());
+
+  for (int i = 0; i < spec.NumInput(); ++i) {
+    dali_proto::InputOutput *in = op.add_input();
+    in->set_name(spec.InputName(i));
+    in->set_device(spec.InputDevice(i));
+    if (spec.IsArgumentInput(i)) {
+        in->set_arg_name(spec.ArgumentInputName(i));
+    }
+    in->set_is_argument_input(spec.IsArgumentInput(i));
+  }
+
+  for (int i = 0; i < spec.NumOutput(); ++i) {
+    dali_proto::InputOutput *out = op.add_output();
+    // Use a placeholder instead of the real name
+    out->set_name(std::to_string(i));
+    out->set_device(spec.OutputDevice(i));
+  }
+
+  auto &schema = spec.GetSchemaOrDefault();
+  std::map<std::string_view, Argument *, std::less<>> sorted_args;
+  for (auto &a : spec.Arguments()) {
+    // Some arguments should be skipped when comparing operators
+    auto arg_name = a->get_name();
+    if (schema.HasArgument(arg_name))
+      if (schema.GetArgument(arg_name).ignore_cmp)
+        continue;
+
+    sorted_args.emplace(arg_name, a.get());
+  }
+
+  for (auto [name, a] : sorted_args) {
+    dali_proto::Argument *arg = op.add_args();
+    DaliProtoPriv arg_wrap(arg);
+    a->SerializeToProtobuf(&arg_wrap);
+  }
+
+  return op.SerializeAsString();
+}
+
+/** The context for Common Subgraph Elimination */
+class CSE {
+ public:
+  void Run(OpGraph &graph) {
+    for (auto &node : graph.OpNodes())
```

Comment:
OpNodes returns the nodes in topological order?

Reply:
Yes, the graph is sorted by the Builder.